### DISCUSSION TITLE
Can't Get Web Search DuckDuckGo Working

### DISCUSSION THREAD

**Discussion by nanafy** [open-webui/open-webui#5191] [General]
# Bug Report

## Installation Method

pip install openwebui
ollama

## Environment

- **Open WebUI Version:** [e.g., v0.3.11]
- **Ollama (if applicable):** [e.g., v0.2.0, v0.1.32-rc1]

- **Operating System:** [e.g., Windows 10, macOS Big Sur, Ubuntu 20.04]
- **Browser (if applicable):** [e.g., Chrome 100.0, Firefox 98.0]

**Confirmation:**

- [Yes] I have read and followed all the instructions provided in the README.md.
- [Yes] I am on the latest version of both Open WebUI and Ollama.
- [Yes] I have included the browser console logs.
- [Yes] I have included the Docker container logs.
- [Yes ] I have provided the exact steps to reproduce the bug in the "Steps to Reproduce" section below.

## Expected Behavior:

DuckDuck go should give results into the output of LLM

## Actual Behavior:
It times out 
INFO:     127.0.0.1:5466 - "POST /api/task/query/completions HTTP/1.1" 400 Bad Request
ERROR [open_webui.apps.rag.main] https://duckduckgo.com RuntimeError: error sending request for url (https://duckduckgo.com/): operation timed out

Caused by:
    operation timed out
Traceback (most recent call last):
  File "path/to/duckduckgo_search.py", line 118, in _get_url
    resp = self.client.request(method, url, params=params, content=content, data=data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: error sending request for url (https://duckduckgo.com/): operation timed out

Caused by:
    operation timed out

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "path/to/main.py", line 919, in store_web_search
    web_results = search_web(
                  ^^^^^^^^^^^
  File "path/to/main.py", line 882, in search_web
    return search_duckduckgo(
           ^^^^^^^^^^^^^^^^^^
  File "path/to/duckduckgo.py", line 27, in search_duckduckgo
    ddgs_gen = ddgs.text(
               ^^^^^^^^^^
  File "path/to/duckduckgo_search.py", line 239, in text
    results = self._text_api(keywords, region, safesearch, timelimit, max_results)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "path/to/duckduckgo_search.py", line 273, in _text_api
    vqd = self._get_vqd(keywords)
          ^^^^^^^^^^^^^^^^^^^^^^^
  File "path/to/duckduckgo_search.py", line 134, in _get_vqd
    resp_content = self._get_url("POST", "https://duckduckgo.com", data={"q": keywords})
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "path/to/duckduckgo_search.py", line 122, in _get_url
    raise TimeoutException(f"{url} {type(ex).__name__}: {ex}") from ex
duckduckgo_search.exceptions.TimeoutException: https://duckduckgo.com RuntimeError: error sending request for url (https://duckduckgo.com/): operation timed out

Caused by:
    operation timed out
INFO:     127.0.0.1:5466 - "POST /rag/api/v1/web/search HTTP/1.1" 400 Bad Request

## Description

**Bug Summary:**
The user has to set concurrent connections and search result count to 1 to see it sort of working. It always fails the first time, but works the second time. If those settings are not set to 1, the context from the web is missed and the llm says does not take into consideration web search results.

## Reproduction Details

Go to admin settings, web browsing, choose duck duck, leave default settings, try to search, it will fail.

## Logs and Screenshots

**Browser Console Logs:**
[Include relevant browser console logs, if applicable]

**Docker Container Logs:**
[Include relevant Docker container logs, if applicable]

**Screenshots/Screen Recordings (if applicable):**
[Attach any relevant screenshots to help illustrate the issue]

## Additional Information

[Include any additional details that may help in understanding and reproducing the issue. This could include specific configurations, error messages, or anything else relevant to the bug.]

## Note

If the bug report is incomplete or does not follow the provided instructions, it may not be addressed. Please ensure that you have followed the steps outlined in the README.md and troubleshooting.md documents, and provide all necessary information for us to reproduce and address the issue. Thank you!

**Comment by snothx** [comment]
macos docker ollama llama llms with openwebui + duckduckgo search enabled in admin settings and search

my duckduckgo search didn't work until i changed "api to auto" inside the docker container

i didn't have nano and vi installed so...

sed -i 's/backend="api"/backend="auto"/g' /app/backend/open_webui/retrieval/web/duckduckgo.py

reboot container and good to go.

**Reply by aaronyanx** [reply]
That doesn't work for me...

here is info:

`INFO:     192.168.215.1:57634 - "GET /static/favicon.png HTTP/1.1" 304 Not Modified
INFO:     192.168.215.1:57650 - "POST /api/v1/chats/0ea5bf56-f8d1-4e8f-8920-7dface1e50c6 HTTP/1.1" 200 OK
INFO:     192.168.215.1:57650 - "GET /api/v1/chats/?page=1 HTTP/1.1" 200 OK
INFO:     192.168.215.1:57650 - "POST /api/v1/memories/query HTTP/1.1" 200 OK
ERROR [open_webui.routers.retrieval] https://lite.duckduckgo.com/lite/ RuntimeError: error sending request for url (https://lite.duckduckgo.com/lite/): operation timed out

Caused by:
    operation timed out
Traceback (most recent call last):
  File "/app/backend/open_webui/routers/retrieval.py", line 1268, in process_web_search
    web_results = search_web(
                  ^^^^^^^^^^^
  File "/app/backend/open_webui/routers/retrieval.py", line 1209, in search_web
    return search_duckduckgo(
           ^^^^^^^^^^^^^^^^^^
  File "/app/backend/open_webui/retrieval/web/duckduckgo.py", line 27, in search_duckduckgo
    ddgs_gen = ddgs.text(
               ^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/duckduckgo_search/duckduckgo_search.py", line 255, in text
    raise DuckDuckGoSearchException(err)
duckduckgo_search.exceptions.DuckDuckGoSearchException: https://lite.duckduckgo.com/lite/ RuntimeError: error sending request for url (https://lite.duckduckgo.com/lite/): operation timed out

Caused by:
    operation timed out
ERROR [open_webui.utils.middleware] 400: https://lite.duckduckgo.com/lite/ RuntimeError: error sending request for url (https://lite.duckduckgo.com/lite/): operation timed out

Caused by:
    operation timed out
Traceback (most recent call last):
  File "/app/backend/open_webui/routers/retrieval.py", line 1268, in process_web_search
    web_results = search_web(
                  ^^^^^^^^^^^
  File "/app/backend/open_webui/routers/retrieval.py", line 1209, in search_web
    return search_duckduckgo(
           ^^^^^^^^^^^^^^^^^^
  File "/app/backend/open_webui/retrieval/web/duckduckgo.py", line 27, in search_duckduckgo
    ddgs_gen = ddgs.text(
               ^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/duckduckgo_search/duckduckgo_search.py", line 255, in text
    raise DuckDuckGoSearchException(err)
duckduckgo_search.exceptions.DuckDuckGoSearchException: https://lite.duckduckgo.com/lite/ RuntimeError: error sending request for url (https://lite.duckduckgo.com/lite/): operation timed out

Caused by:
    operation timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/app/backend/open_webui/utils/middleware.py", line 429, in chat_web_search_handler
    results = await loop.run_in_executor(
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/backend/open_webui/utils/middleware.py", line 431, in <lambda>
    lambda: process_web_search(
            ^^^^^^^^^^^^^^^^^^^
  File "/app/backend/open_webui/routers/retrieval.py", line 1274, in process_web_search
    raise HTTPException(
fastapi.exceptions.HTTPException: 400: https://lite.duckduckgo.com/lite/ RuntimeError: error sending request for url (https://lite.duckduckgo.com/lite/): operation timed out

Caused by:
    operation timed out
INFO:     192.168.215.1:57650 - "POST /api/chat/completions HTTP/1.1" 200 OK
INFO:     192.168.215.1:57650 - "GET /api/v1/chats/?page=1 HTTP/1.1" 200 OK
INFO:     192.168.215.1:57622 - "POST /api/v1/tasks/auto/completions HTTP/1.1" 200 OK
`

Could you help me please..

**Comment by kylespencer** [comment]
In recent versions of OpenWebUI I get Rate Limit errors when using DuckDuckGo as the Web Search back end. This happens on both Docker and UVX based instances, and in different hosting environments with different IP space.

**Reply by ep2610** [reply]
Is there any update on this?

**Comment by digitalextremist** [comment]
Seems like `duckduckgo` search stopped working for me around this point:  https://github.com/open-webui/open-webui/commit/760ea3f4afbbaa4f2046dc53aa9ea3fd2f784c53

Which references this: https://github.com/deedy5/duckduckgo_search/commit/3ee8e08b1c8efc9b7356f406db299f1114dc4860

**Comment by i5hi** [comment]
i keep getting:
```
An error occurred while searching the web
```

When trying to use duckduckgo for Web Search

**Comment by Aillian** [comment]
same here

**Reply by rgaricano** [reply]
version? it was fixed in v.0.6.17

**Comment by davidshen84** [comment]
Has DuckDuckGo search been removed? On the Web Search page of the admin panel, I cannot find the 'duckduckgo' search engine. I did find an engine called 'DDGS', but when I chose it and performed a query with web search enabled, I got an error saying timeout connecting to bing.com.

```
open-webui   File "/usr/local/lib/python3.11/site-packages/ddgs/ddgs.py", line 190, in text
open-webui     raise DDGSException(err)
open-webui           │             └ TimeoutException('https://www.bing.com/search RuntimeError: error sending request for url (https://www.bing.com/search?q=flig...
open-webui           └ <class 'ddgs.exceptions.DDGSException'>
open-webui 
open-webui ddgs.exceptions.DDGSException: https://www.bing.com/search RuntimeError: error sending request for url (https://www.bing.com/search?q=11111): operation timed out
```
The error is from a component named `DDGS`, but it is trying to use the *Bing* search engine. But there's a separate search engine called 'bing' on that page.

**Reply by rgaricano** [reply]
DDGS begun as duckduckgo search lib but now is a metasearch lib, I commented it in different threads:
https://github.com/open-webui/open-webui/issues/16080#issuecomment-3126006541
https://github.com/open-webui/open-webui/pull/16694
https://github.com/open-webui/open-webui/issues/16630#issuecomment-3189277109
...

**Comment by davidshen84** [comment]
It is not documented at all? The closest I can find is this:
https://docs.openwebui.com/tutorials/web-search/external#example-implementation-pythonfastapi

**Reply by rgaricano** [reply]
https://github.com/open-webui/open-webui/blob/6bc5d331a27c5106f492213510a763effa316faf/CHANGELOG.md?plain=1#L243

**Comment by davidshen84** [comment]
Thanks.

But that line doesn't tell us how to configure DDGS in OpenWebUI. On
https://github.com/deedy5/ddgs, it also doesn't suggest that it can read
some environment variables to change its behaviour.

If I choose DDGS, what else should I do to use the duckduckgo search engine?

**Reply by Classic298** [reply]
PR for the /docs repository welcome!
Feel free to discuss it there since obviously a tutorial for DDGS can be added there.

**Comment by davidshen84** [comment]
https://github.com/open-webui/open-webui/blob/main/backend/open_webui/retrieval/web/duckduckgo.py#L37

Looking at this line, it is hardcoded to use a backend named "lite", which
is not defined on the DDGS website.

**Reply by rgaricano** [reply]
Yes, but it was 2 months ago, before they changed the lib to a metasearch engine:
https://github.com/deedy5/ddgs/blame/dd3015638ece8df17f5997ee3d9b703d9ee5bdc2/README.md#L225-L235
```
    Args:
        keywords: keywords for query.
        region: us-en, uk-en, ru-ru, etc. Defaults to None.
        safesearch: on, moderate, off. Defaults to "moderate".
        timelimit: d, w, m, y. Defaults to None.
        backend: auto, html, lite. Defaults to auto.
            auto - try all backends in random order,
            html - collect data from https://html.duckduckgo.com,
            lite - collect data from https://lite.duckduckgo.com,
            bing - collect data from https://www.bing.com.
        max_results: max number of results. If None, returns results only from the first response. Defaults to None.
```

And yes, a PR is necessary to be able to configure the new functionalities of this library.

**Comment by ptim-miller** [comment]
I wasn't able to get ddgs to work with the current version (v0.6.34) until I forced an update (pip3 install ddgs==9.6.1). The required version established in backend/requirements.txt is 9.0.0. Further testing may be needed to deconflict, but it seems to work without any other code changes.

**Reply by Rotios** [reply]
I managed to update to [9.9.0](https://pypi.org/project/ddgs/#history) which is the latest version. 

For those who are wondering how to do this when running an official docker container, you can follow these steps:

1. Get the docker container id using `docker ps`
```
> docker ps
CONTAINER ID   IMAGE          COMMAND           CREATED       STATUS                    PORTS                                         NAMES
4c53a9870c82   0e2db1ae86ce   "bash start.sh"   11 days ago   Up 12 minutes (healthy)   0.0.0.0:3000->8080/tcp, [::]:3000->8080/tcp   open-webui
```
2. Enter the container using `docker exec`
```
> docker exec -it 4c53a9870c82   /bin/sh
#
```
3. (Optional) Check version of ddgs using `pip show ddgs`
```
# pip show ddgs
Name: ddgs
Version: 9.0.0
Summary: Dux Distributed Global Search. A metasearch library that aggregates results from diverse web search services.
Home-page:
Author: deedy5
Author-email:
License: MIT License
Location: /usr/local/lib/python3.11/site-packages
Requires: click, lxml, primp
Required-by:
```
4. Update the requirements.txt file with the latest version of ddgs. 
```
# sed -i 's/^ddgs==.*$/ddgs==9.9.0/' requirements.txt
```
5. Verify the update succeeded
```
# cat requirements.txt | grep ddgs
ddgs==9.9.0
```
6. Run pip install on the update requirements.txt file
```
# pip install -r requirements.txt
```
7. Validate latest version of ddgs is installed using `pip show ddgs`
```
# pip show ddgs
Name: ddgs
Version: 9.9.0
Summary: Dux Distributed Global Search. A metasearch library that aggregates results from diverse web search services.
Home-page: https://github.com/deedy5/ddgs
Author: deedy5
Author-email:
License-Expression: MIT
Location: /usr/local/lib/python3.11/site-packages
Requires: click, httpx, lxml, primp
Required-by:
```

You should not have to do anything else. OpenWebUI should pick up the latest version of ddgs automatically in this container. 

If you run a fresh container or switch to a newer image, you will have to perform these actions again.

**Reply by RyanHope** [reply]
The described fix above works for the uv based install as well. I activated the venv, used ensurepip to get pip, then upgraded ddgs. After that DDGS worked.

**Reply by neuhaus** [reply]
my ddgs isn't working when I configure it to use Duckduckgo.
I checked the version inside the container and it's already version 9.10.0.

I don't want results from Grokipedia, who thought that was a good idea to include it in the list of sources?

**Comment by MidKnightXI** [comment]
Same issue for me here in the docker based openwebui, with DDGS and duckduckgo search engine, even after upgrading DDGS to the last version available inside the container.
